sdk/test-utils/recorder/src/utils.ts

```diff
@@ -21,17 +21,80 @@ export function isPlaybackMode() {
   return !isRecordMode() && !isLiveMode();
 }
 
+/**
+ * Encodes a string as a URI component, but also taking in consideration the RFC 3986 specification.
+ * JavaScript's encodeURIComponent method doesn't take in consideration the characters: !, ', (, ), *
+ * @param str The string that needs to be encoded.
+ */
+export function encodeRFC3986(str: string): string {
+  return encodeURIComponent(str).replace(
+    /[!'()*]/g,
+    (x) =>
+      `%${x
+        .charCodeAt(0)
+        .toString(16)
+        .toUpperCase()}`
+  );
+}
+
+/**
+ * Escapes all of the valid RegExp characters of a string.
+ * @param str The string that needs to be escaped.
+ */
 export function escapeRegExp(str: string): string {
-  return encodeURIComponent(str)
-    .replace(
-      /[!'()*]/g,
-      (x) =>
-        `%${x
-          .charCodeAt(0)
-          .toString(16)
-          .toUpperCase()}`
-    ) // RFC 3986.
-    .replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&"); // All the RegExp sensitive characters.
+  return str.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
+}
+
+/**
+ * Replaces all occurrences of a pattern in a string with a given replacement.
+ * @param string Target of the replacements.
+ * @param pattern String used to match and find what to replace.
+ * @param replacement Replacement of the matched string.
+ */
+function replaceAll(string: string, pattern: string, replacement: string) {
+  return string.replace(new RegExp(escapeRegExp(pattern), "g"), replacement);
+}
+
+export type ReplacementDictionary = { [x: string]: string };
+
+/**
+ * Looks for the environment variables based on the keys of the given dictionary,
+ * then replaces the values found with each value from the same dictionary.
+ * @param replacements A dictionary of string keys and string values.
+ * @param content The content that has the text to be replaced.
+ */
+export function applyReplacementDictionary(
+  env: NodeJS.ProcessEnv,
```

Comment:
I haven't looked at the complete PR yet. I'll do another pass.

At first glance, I see a problem.
- `applyReplacementDictionary` is being used in the `filterSecrets` method
-  `filterSecrets` method is a member of `BaseRecorder` class
- `BaseRecorder` class is meant for both browser and node recorders
- Which means..  `filterSecrets` method is meant for both node and browser tests
- So, you shouldn't do `NodeJS.ProcessEnv` in the browsers since that would break the browser tests in record/playback modes.

`env` elsewhere is globally defined as 
> export const env = isBrowser() ? (window as any).__env__ : process.env;

So, technically, you don't even have to pass the env.
My suggestion is to remove the argument `env` from here and wherever this function is being called, everything else will adjust by itself IMO.

Reply:
NodeJS.ProcessEnv is just a type, not an actual value. Is that still a problem?

Reply:
I just tested this code with KeyVault-Keys browser tests. They passed just fine.

Reply:
I was wondering if this might cause issues.

> I just tested this code with KeyVault-Keys browser tests. They passed just fine.

How did you verify? Have you recorded the tests and then played back?

Reply:
Yep, I did record the tests and then played them back.

I think this works fine because NodeJS.ProcessEnv is just a type, not a real value. The types are agnostic of the environment.